### PR TITLE
Add api/v2 with GRPC + REST

### PR DESCRIPTION
**Pull request by fabxc**
@Gouthamve @brian-brazil 

Not that the new delete endpoint is not a GET anymore with our textual label matcher representation. It's a POST request, which seems a bit safer, and you just have to provide the matchers as proper JSON. Seems fine for the use case.

Example:
```
curl -XPOST http://localhost:9090/api/v2/admin/tsdb/delete_series -H "Content-Type: application/json" -d'
{"mint":"2017-05-08T10:25:08Z","maxt":"2017-08-08T10:25:08Z","matchers": [{"type":"EQ","name":"__name__","value":"prometheus_notifications_queue_capacity"}]}'
```

Also in this case we just accept RFC3339 timestamps. I lean towards sticking with a single representation once we extend v2 for queries as well.

### CODE REVIEW

**Comment by roman-vynar** [comment]
Any reason why not DELETE instead of POST to some kind of the list endpoint?

**Comment by fabxc** [comment]
@roman-vynar DELETE has no body, which means we'd have to encode the non-trivial query in url parameters for the REST call. In the v1 API we use a PromQL-like selector as the URL parameters. I suppose it's still possible, but I wasn't sure that it's super relevant in an admin API.

Happy to reconsider and feedback from more people would certainly help here.

**Comment by Gouthamve** [comment]
Can we not have a simple query for selection instead of all the intricate details:

```
{
  "match": "{__name__=\"yolo\", path=~\"/api/.*\"}"
}
```

**Comment by fabxc** [comment]
We could, but I'd like to keep at least our proto API well structured so people coding against it don't have to generate the correct selector language on top of gRPC.
Similar arguments could be made for the REST API as well, though that one profits more from something easier for humans to type. Then again, it's a low-throughput admin API.

**Comment by fabxc** [comment]
Updated, tests all passing now.

@brian-brazil opinions on the delete API endpoint selector?

File: web/api/v2/v2pb/api.proto [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_getters_all) = false;
+
+// AdminTSDB exposes methods of Prometheus' TSDB.
+service AdminTSDB {
+  // Reload triggers a reload of block data.
+  rpc Reload(AdminTSDBReloadRequest) returns (AdminTSDBReloadResponse) {
+    option (google.api.http) = {
+      get: "/admin/tsdb/reload"
```
Comment (brian-brazil) [inline web/api/v2/v2pb/api.proto]: This doesn't do anything to the tsdb

Comment (fabxc) [inline web/api/v2/v2pb/api.proto]: Yes, hence the handler says "not implemented". We can extend the API incrementally once we've settled on the basics here.

**Comment by tomwilkie** [comment]
> I'd like to keep at least our proto API well structured so people coding against it don't have to generate the correct selector language on top of gRPC.

+1 - even better would be to use the same proto definitions as the remote query API: https://github.com/prometheus/prometheus/blob/master/storage/remote/remote.proto#L60

**Comment by fabxc** [comment]
Yea, I've been thinking about in how far proto dependencies should be used. It would certainly be odd to import from some remote-proto to serve a regular API. So the LabelSelector message should reasonably go into its own, independent package. There I have my concerns that we end up with a rigid structure that's prohibitive of use-case specific customizations, just like with our common/model repository.

A little duplication for effectively independent APIs doesn't seem too bad.

**Comment by tomwilkie** [comment]
Agreed, I wouldn't put this in a different repo but a shared package in
Prometheus repo might be useful - especially as the functions to convert to
and from native types can be shared.

**Comment by fabxc** [comment]
Can we safely move the remote proto to another file without breaking wire compatibility?
Question is where to, I suspect `pkg/labels/labelspb` would be suitable for `LabelSelector` and `LabelPair`?

Shame that the generated pb exposes those obscure error variables. Otherwise I'd be more inclined to skip the separate proto packages.

**Comment by brian-brazil** [comment]
> Can we safely move the remote proto to another file without breaking wire compatibility?

Yes. As long as the field numbers and types remain the same, we're all good.

> Question is where to, I suspect pkg/labels/labelspb would be suitable for LabelSelector and LabelPair?

A single directory with all our public protos is probably best.

File: web/api/v2/v2pb/api.proto [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+message AdminTSDBReloadRequest {
+}
+
+message AdminTSDBReloadResponse {
+}
+
+message AdminTSDBSeriesDeleteRequest {
+  google.protobuf.Timestamp mint = 1 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
+  google.protobuf.Timestamp maxt = 2 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
+  repeated LabelMatcher matchers = 3;
```
Comment (xiang90) [inline web/api/v2/v2pb/api.proto]: we probably need to document this field.

Comment (xiang90) [inline web/api/v2/v2pb/api.proto]: the series will be deleted if there exists a match? or all matches must be matched?

File: web/api/v2/v2pb/api.proto [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  enum Type {
+    EQ  = 0;
+    NEQ = 1;
+    RE  = 2;
+    NRE = 3;
+  }
+  Type type    = 1;
+  string name  = 2;
+  string value = 3;
+}
```
Comment (xiang90) [inline web/api/v2/v2pb/api.proto]: new line here.

File: web/web.go [hunk 4]
```diff
@@ -30,15 +30,21 @@ import (
 	"sync"
 	"time"
 
+	"google.golang.org/grpc"
```
Comment (xiang90) [inline web/web.go]: move this import line to the block starting from line38?

File: web/web.go [hunk 5]
```diff
@@ -256,24 +271,51 @@ func (h *Handler) Reload() <-chan chan error {
 }
 
 // Run serves the HTTP endpoints.
-func (h *Handler) Run() {
+func (h *Handler) Run(ctx context.Context) error {
```
Comment (xiang90) [inline web/web.go]: where do you use the passed in ctx?

**Comment by tomwilkie** [comment]
>> Can we safely move the remote proto to another file without breaking wire compatibility?
> Yes. As long as the field numbers and types remain the same, we're all good.

Doesn't apply in this case as remote.proto doesn't describe any gRPC services, but for gRPC the package name is included in the service name.

**Pull request closed without merging**